File: my_vpn/app/src/main/java/com/myvpn/simple/ClipboardHelper.java
package com.myvpn.simple;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import java.util.List;
import java.util.ArrayList;

public class ClipboardHelper {
    
    public static List<TrojanConfig> readSubscriptionFromClipboard(Context context) {
        ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        
        if (clipboard != null && clipboard.hasPrimaryClip()) {
            ClipData clipData = clipboard.getPrimaryClip();
            if (clipData != null && clipData.getItemCount() > 0) {
                CharSequence text = clipData.getItemAt(0).getText();
                if (text != null) {
                    String clipboardText = text.toString().trim();
                    
                    if (clipboardText.startsWith("trojan://")) {
                        TrojanConfig config = SubscriptionParser.parseTrojanUrl(clipboardText);
                        if (config != null && config.isValid()) {
                            List<TrojanConfig> result = new ArrayList<>();
                            result.add(config);
                            return result;
                        }
                    } else if (isBase64(clipboardText)) {
                        return SubscriptionParser.parseSubscription(clipboardText);
                    }
                }
            }
        }
        
        return new ArrayList<>();
    }
    
    private static boolean isBase64(String str) {
        try {
            String decoded = new String(android.util.Base64.decode(str, android.util.Base64.DEFAULT));
            return decoded.contains("trojan://");
        } catch (Exception e) {
            return false;
        }
    }
}

File: kt/app/src/main/kotlin/com/myvpn/simple/ClipboardHelper.kt
package com.myvpn.simple

import android.content.ClipboardManager
import android.content.Context
import android.util.Base64
import com.myvpn.simple.utils.SubscriptionDownloader

object ClipboardHelper {

    fun readSubscriptionFromClipboard(context: Context): List<TrojanConfig> {
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
            ?: return emptyList()

        if (!clipboard.hasPrimaryClip()) {
            return emptyList()
        }

        val clipData = clipboard.primaryClip ?: return emptyList()
        if (clipData.itemCount == 0) {
            return emptyList()
        }

        val text = clipData.getItemAt(0).text?.toString()?.trim() ?: return emptyList()

        return when {
            // 单个 Trojan URL
            text.startsWith("trojan://") -> {
                val config = SubscriptionParser.parseTrojanUrl(text)
                if (config != null && config.isValid()) {
                    listOf(config)
                } else {
                    emptyList()
                }
            }
            // 订阅链接 URL
            SubscriptionDownloader.isSubscriptionUrl(text) -> {
                val content = SubscriptionDownloader.downloadSubscription(text)
                if (content != null) {
                    SubscriptionParser.parseSubscription(content)
                } else {
                    emptyList()
                }
            }
            // Base64 编码的订阅内容
            isBase64(text) -> {
                SubscriptionParser.parseSubscription(text)
            }
            else -> emptyList()
        }
    }

    private fun isBase64(str: String): Boolean {
        return try {
            val decoded = String(Base64.decode(str, Base64.DEFAULT))
            decoded.contains("trojan://")
        } catch (e: Exception) {
            false
        }
    }
}
